File: motor3d 2.0/octree.vb
Imports Motor3D.Primitivas3D

Namespace Espacio3D
    Public Class Octree
        Inherits ObjetoGeometrico3D

        Private mNiveles As Integer

        Private mSectorRaiz As SectorOctree

        Public Property Espacio() As AABB3D
            Get
                Return mSectorRaiz.Espacio
            End Get
            Set(ByVal value As AABB3D)
                If mSectorRaiz.Espacio <> value Then
                    mSectorRaiz = New SectorOctree(mNiveles, value)
                End If
            End Set
        End Property

        Public Property Niveles() As Integer
            Get
                Return mNiveles
            End Get
            Set(ByVal value As Integer)
                If value > 0 Then
                    mNiveles = value
                    mSectorRaiz = New SectorOctree(mNiveles, mSectorRaiz.Espacio)
                End If
            End Set
        End Property

        Public ReadOnly Property SectorRaiz() As SectorOctree
            Get
                Return mSectorRaiz
            End Get
        End Property

        Public ReadOnly Property Pertenece(ByVal Punto As Punto3D) As Boolean
            Get
                Return mSectorRaiz.Espacio.Pertenece(Punto)
            End Get
        End Property

        Public ReadOnly Property Pertenece(ByVal AABB As AABB3D) As Boolean
            Get
                Return mSectorRaiz.Espacio.Colisionan(AABB)
            End Get
        End Property

        Public ReadOnly Property Pertenece(ByVal Poliedro As Poliedro, Optional ByVal AABBSUR As Boolean = False) As Boolean
            Get
                If AABBSUR Then
                    Return mSectorRaiz.Espacio.Colisionan(Poliedro.AABBSUR)
                Else
                    Return mSectorRaiz.Espacio.Colisionan(Poliedro.AABBSRC)
                End If
            End Get
        End Property

        Public ReadOnly Property Pertenece(ByVal Recta As Recta3D) As Boolean
            Get
                Return mSectorRaiz.Espacio.Pertenece(Recta)
            End Get
        End Property

        Public Sub New(ByVal Niveles As Integer, ByVal AABB As AABB3D)
            If Niveles > 1 Then
                mNiveles = Niveles

                mSectorRaiz = New SectorOctree(Niveles, AABB)
            Else
                Throw New ExcepcionGeometrica3D("OCTREE (NEW): Un octree debe tener al menos dos niveles:" & vbNewLine _
                                                & "Niveles=" & Niveles)
            End If
        End Sub

        Public Sub Refrescar()
            mSectorRaiz.Refrescar()
        End Sub

        Public Function Sector(ByVal Punto As Punto3D) As SectorOctree
            Return Sector(Me, Punto)
        End Function

        Public Function Sector(ByVal Recta As Recta3D) As SectorOctree
            Return Sector(Me, Recta)
        End Function

        Public Function Sector(ByVal AABB As AABB3D) As SectorOctree
            Return Sector(Me, AABB)
        End Function

        Public Function Sector(ByVal Poliedro As Poliedro, Optional ByVal AABBSUR As Boolean = False) As SectorOctree
            Return Sector(Me, Poliedro, AABBSUR)
        End Function

        Public Function Sectores(ByVal Recta As Recta3D) As SectorOctree()
            Return Sectores(Me, Recta)
        End Function

        Public Shared Function Sector(ByVal Octree As Octree, ByVal Punto As Punto3D) As SectorOctree
            Dim Resultado As Integer = Octree.SectorRaiz.Pertenece(Punto)
            Dim S As SectorOctree = Octree.SectorRaiz

            If Resultado <> -2 Then
                If Resultado = -1 Then
                    Return S
                Else
                    Do While Resultado <> -2 AndAlso Resultado <> -1
                        Resultado = S.Pertenece(Punto)
                        If Resultado <> -2 Then
                            If Resultado = -1 Then
                                Return S
                            Else
                                S = S.Hijos(Resultado)
                            End If
                        Else
                            Return S.Padre
                        End If
                    Loop

                    Return S
                End If
            Else
                Throw New ExcepcionGeometrica3D("OCTREE (PERTENECE): El punto especificado no pertenece al espacio dominado por el quadtree." & vbNewLine _
                                                & "Punto=" & Punto.ToString & vbNewLine _
                                                & "Espacio=" & Octree.Espacio.ToString)
            End If
        End Function

        Public Shared Function Sector(ByVal Octree As Octree, ByVal Recta As Recta3D) As SectorOctree
            Dim Resultado As Integer = Octree.SectorRaiz.Pertenece(Recta)
            Dim S As SectorOctree = Octree.SectorRaiz

            If Resultado <> -2 Then
                If Resultado = -1 Then
                    Return S
                Else
                    Do While Resultado <> -2 AndAlso Resultado <> -1
                        Resultado = S.Pertenece(Recta)
                        If Resultado <> -2 Then
                            If Resultado = -1 Then
                                Return S
                            Else
                                S = S.Hijos(Resultado)
                            End If
                        Else
                            Return S.Padre
                        End If
                    Loop

                    Return S
                End If
            Else
                Throw New ExcepcionGeometrica3D("OCTREE (PERTENECE): La recta especificada no pertenece al espacio dominado por el quadtree." & vbNewLine _
                                                & "Punto=" & Recta.ToString & vbNewLine _
                                                & "Espacio=" & Octree.Espacio.ToString)
            End If
        End Function

        Public Shared Function Sectores(ByVal Octree As Octree, ByVal Recta As Recta3D) As SectorOctree()
            Dim S As SectorOctree = Octree.SectorRaiz
            Dim Retorno As New List(Of SectorOctree)

            If Octree.SectorRaiz.Pertenece(Recta) Then
                InterseccionRecta(Recta, S, Retorno)

                Return Retorno.ToArray
            Else
                Throw New ExcepcionGeometrica3D("OCTREE (PERTENECE): La recta especificado no pertenece al espacio dominado por el quadtree." & vbNewLine _
                                                & "Recta=" & Recta.ToString & vbNewLine _
                                                & "Espacio=" & Octree.Espacio.ToString)
            End If
        End Function

        Public Shared Sub InterseccionRecta(ByVal Recta As Recta3D, ByVal Sector As SectorOctree, ByRef ListaRetorno As List(Of SectorOctree))
            If Not Sector.EsHoja Then
                For Each Hijo As SectorOctree In Sector.Hijos
                    If Hijo.Pertenece(Recta) Then
                        InterseccionRecta(Recta, Hijo, ListaRetorno)
                    End If
                Next
            Else
                If Sector.Pertenece(Recta) Then ListaRetorno.Add(Sector)
            End If
        End Sub

        Public Shared Function Sector(ByVal Octree As Octree, ByVal AABB As AABB3D) As SectorOctree
            Dim Resultado As Integer = Octree.SectorRaiz.Pertenece(AABB)
            Dim S As SectorOctree = Octree.SectorRaiz

            'PARA ENTENDER EL ALGORITMO, IR A FUNCION PERTENECE DE SECTORQUADTREE.
            If Resultado <> -2 Then
                If Resultado = -1 Then
                    Return Octree.SectorRaiz
                Else
                    Do
                        If Resultado <> -2 Then
                            If Resultado = -1 Then
                                Return S
                            Else
                                S = S.Hijos(Resultado)
                            End If
                        Else
                            Return S.Padre
                        End If
                        Resultado = S.Pertenece(AABB)
                    Loop While Resultado <> -2 AndAlso Resultado <> -1

                    Return S
                End If
            Else
                Throw New ExcepcionGeometrica3D("OCTREE (PERTENECE): La AABB especificada no pertenece al espacio dominado por el quadtree." & vbNewLine _
                                                & "AABB=" & AABB.ToString & vbNewLine _
                                                & "Espacio=" & Octree.Espacio.ToString)
            End If
        End Function

        Public Shared Function Sector(ByVal Octree As Octree, ByRef Poliedro As Poliedro, Optional ByVal AABBSUR As Boolean = False) As SectorOctree
            Dim Resultado As Integer = Octree.SectorRaiz.Pertenece(Poliedro, AABBSUR)
            Dim S As SectorOctree = Octree.SectorRaiz

            'PARA ENTENDER EL ALGORITMO, IR A FUNCION PERTENECE DE SECTORQUADTREE.
            If Resultado <> -2 Then
                If Resultado = -1 Then
                    Return Octree.SectorRaiz
                Else
                    Do
                        If Resultado <> -2 Then
                            If Resultado = -1 Then
                                Return S
                            Else
                                S = S.Hijos(Resultado)
                            End If
                        Else
                            Return S.Padre
                        End If
                        Resultado = S.Pertenece(Poliedro, AABBSUR)
                    Loop While Resultado <> -2 AndAlso Resultado <> -1

                    Return S
                End If
            Else
                Throw New ExcepcionGeometrica3D("OCTREE (PERTENECE): El poliedro especificado no pertenece al espacio dominado por el quadtree." & vbNewLine _
                                                & "Poliedro=" & Poliedro.ToString & vbNewLine _
                                                & "Espacio=" & Octree.Espacio.ToString)
            End If
        End Function

        Public Overrides Function ToString() As String
            Return "{Octree de espacio=" & mSectorRaiz.Espacio.ToString & "}"
        End Function
    End Class
End Namespace

File: motor3d 2.0.ConvertedToC#/Octree.cs
using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
using System.Linq;
using System.Xml.Linq;
using Motor3D.Primitivas3D;

namespace Motor3D.Espacio3D
{
	public class Octree : ObjetoGeometrico3D
	{


		private int mNiveles;

		private SectorOctree mSectorRaiz;
		public Caja3D Espacio {
			get { return mSectorRaiz.Espacio; }
			set {
				if (mSectorRaiz.Espacio != value) {
					mSectorRaiz = new SectorOctree(mNiveles, value);
				}
			}
		}

		public int Niveles {
			get { return mNiveles; }
			set {
				if (value > 0) {
					mNiveles = value;
					mSectorRaiz = new SectorOctree(mNiveles, mSectorRaiz.Espacio);
				}
			}
		}

		public SectorOctree SectorRaiz {
			get { return mSectorRaiz; }
		}

		public bool Pertenece {
			get { return mSectorRaiz.Espacio.Pertenece(Punto); }
		}

		public bool Pertenece {
			get { return mSectorRaiz.Espacio.Colisionan(Caja); }
		}

		public bool Pertenece {
			get {
				if (CajaSUR) {
					return mSectorRaiz.Espacio.Colisionan(Poliedro.CajaSUR);
				} else {
					return mSectorRaiz.Espacio.Colisionan(Poliedro.CajaSRC);
				}
			}
		}

		public bool Pertenece {
			get { return mSectorRaiz.Espacio.Pertenece(Recta); }
		}

		public Octree(int Niveles, Caja3D Caja)
		{
			if (Niveles > 1) {
				mNiveles = Niveles;

				mSectorRaiz = new SectorOctree(Niveles, Caja);
			} else {
				throw new ExcepcionGeometrica3D("OCTREE (NEW): Un octree debe tener al menos dos niveles:" + Constants.vbNewLine + "Niveles=" + Niveles);
			}
		}

		public void Refrescar()
		{
			mSectorRaiz.Refrescar();
		}

		public SectorOctree Sector(Punto3D Punto)
		{
			return Sector(this, Punto);
		}

		public SectorOctree Sector(Recta3D Recta)
		{
			return Sector(this, Recta);
		}

		public SectorOctree Sector(Caja3D Caja)
		{
			return Sector(this, Caja);
		}

		public SectorOctree Sector(Poliedro Poliedro, bool CajaSUR = false)
		{
			return Sector(this, ref Poliedro, CajaSUR);
		}

		public SectorOctree[] Sectores(Recta3D Recta)
		{
			return Sectores(this, Recta);
		}

		public static SectorOctree Sector(Octree Octree, Punto3D Punto)
		{
			int Resultado = Octree.SectorRaiz.Pertenece(ref Punto);
			SectorOctree S = Octree.SectorRaiz;

			if (Resultado != -2) {
				if (Resultado == -1) {
					return S;
				} else {
					while (Resultado != -2 && Resultado != -1) {
						Resultado = S.Pertenece(ref Punto);
						if (Resultado != -2) {
							if (Resultado == -1) {
								return S;
							} else {
								S = S.Hijos[Resultado];
							}
						} else {
							return S.Padre;
						}
					}

					return S;
				}
			} else {
				throw new ExcepcionGeometrica3D("OCTREE (PERTENECE): El punto especificado no pertenece al espacio dominado por el quadtree." + Constants.vbNewLine + "Punto=" + Punto.ToString() + Constants.vbNewLine + "Espacio=" + Octree.Espacio.ToString());
			}
		}

		public static SectorOctree Sector(Octree Octree, Recta3D Recta)
		{
			int Resultado = Octree.SectorRaiz.Pertenece(ref Recta);
			SectorOctree S = Octree.SectorRaiz;

			if (Resultado != -2) {
				if (Resultado == -1) {
					return S;
				} else {
					while (Resultado != -2 && Resultado != -1) {
						Resultado = S.Pertenece(ref Recta);
						if (Resultado != -2) {
							if (Resultado == -1) {
								return S;
							} else {
								S = S.Hijos[Resultado];
							}
						} else {
							return S.Padre;
						}
					}

					return S;
				}
			} else {
				throw new ExcepcionGeometrica3D("OCTREE (PERTENECE): La recta especificada no pertenece al espacio dominado por el quadtree." + Constants.vbNewLine + "Punto=" + Recta.ToString() + Constants.vbNewLine + "Espacio=" + Octree.Espacio.ToString());
			}
		}

		public static SectorOctree[] Sectores(Octree Octree, Recta3D Recta)
		{
			SectorOctree S = Octree.SectorRaiz;
			List<SectorOctree> Retorno = new List<SectorOctree>();

			if (Octree.SectorRaiz.Pertenece(ref Recta)) {
				InterseccionRecta(Recta, S, ref Retorno);

				return Retorno.ToArray();
			} else {
				throw new ExcepcionGeometrica3D("OCTREE (PERTENECE): La recta especificado no pertenece al espacio dominado por el quadtree." + Constants.vbNewLine + "Recta=" + Recta.ToString() + Constants.vbNewLine + "Espacio=" + Octree.Espacio.ToString());
			}
		}

		public static void InterseccionRecta(Recta3D Recta, SectorOctree Sector, ref List<SectorOctree> ListaRetorno)
		{
			if (!Sector.EsHoja) {
				foreach (SectorOctree Hijo in Sector.Hijos) {
					if (Hijo.Pertenece(ref Recta)) {
						InterseccionRecta(Recta, Hijo, ref ListaRetorno);
					}
				}
			} else {
				if (Sector.Pertenece(ref Recta))
					ListaRetorno.Add(Sector);
			}
		}

		public static SectorOctree Sector(Octree Octree, Caja3D Caja)
		{
			int Resultado = Octree.SectorRaiz.Pertenece(ref Caja);
			SectorOctree S = Octree.SectorRaiz;

			//PARA ENTENDER EL ALGORITMO, IR A FUNCION PERTENECE DE SECTORQUADTREE.
			if (Resultado != -2) {
				if (Resultado == -1) {
					return Octree.SectorRaiz;
				} else {
					do {
						if (Resultado != -2) {
							if (Resultado == -1) {
								return S;
							} else {
								S = S.Hijos[Resultado];
							}
						} else {
							return S.Padre;
						}
						Resultado = S.Pertenece(ref Caja);
					} while (Resultado != -2 && Resultado != -1);

					return S;
				}
			} else {
				throw new ExcepcionGeometrica3D("OCTREE (PERTENECE): La caja especificada no pertenece al espacio dominado por el quadtree." + Constants.vbNewLine + "Caja=" + Caja.ToString() + Constants.vbNewLine + "Espacio=" + Octree.Espacio.ToString());
			}
		}

		public static SectorOctree Sector(Octree Octree, ref Poliedro Poliedro, bool CajaSUR = false)
		{
			int Resultado = Octree.SectorRaiz.Pertenece(ref Poliedro, CajaSUR);
			SectorOctree S = Octree.SectorRaiz;

			//PARA ENTENDER EL ALGORITMO, IR A FUNCION PERTENECE DE SECTORQUADTREE.
			if (Resultado != -2) {
				if (Resultado == -1) {
					return Octree.SectorRaiz;
				} else {
					do {
						if (Resultado != -2) {
							if (Resultado == -1) {
								return S;
							} else {
								S = S.Hijos[Resultado];
							}
						} else {
							return S.Padre;
						}
						Resultado = S.Pertenece(ref Poliedro, CajaSUR);
					} while (Resultado != -2 && Resultado != -1);

					return S;
				}
			} else {
				throw new ExcepcionGeometrica3D("OCTREE (PERTENECE): El poliedro especificado no pertenece al espacio dominado por el quadtree." + Constants.vbNewLine + "Poliedro=" + Poliedro.ToString() + Constants.vbNewLine + "Espacio=" + Octree.Espacio.ToString());
			}
		}

		public override string ToString()
		{
			return "{Octree de espacio=" + mSectorRaiz.Espacio.ToString() + "}";
		}
	}
}


